### PR TITLE
Bybit USDT Perpetual

### PR DESCRIPTION
**Pull request by trendoscope-algorithms**
Bybit Futures API for USD Perpetual - completed parts where online trading can be done. There are still some unimplemented methods. Will check when free.

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4456"></a>

### CODE REVIEW

**Comment by dylburger** [comment]
Thank you so much for submitting this! We've added it to our backlog to review, and our team has been notified.

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

**Comment by trendoscope-algorithms** [comment]
Few tests.

image

image

image

image

image

image

image

image

image

image

Done all the testing. Too lazy to get screenshots for each. Leave that to you :)

**Comment by feyzullah** [comment]
Hi @trendoscope-algorithms, can you please rename files and folders as `slug-component-name` to `component-name`? e.g. `bybit-account-add-reduce-margin` to `account-add-reduce-margin`. Component keys are ok and should be left as the same. I remember that in your last PR it was the same but I was confused there too, sorry for that.

File: components/bybit/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@pipedream/bybit",
-  "version": "0.0.1",
+  "version": "0.0.2",
   "description": "Pipedream ByBit Components",
   "main": "dist/app/bybit.app.mjs",
```
Comment (feyzullah) [inline components/bybit/package.json:5]:
```suggestion
  "main": "bybit.app.mjs",
```

File: components/bybit/actions/linear-market-query-kline/linear-market-query-kline.mjs [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import bybit from "../../bybit.app.mjs";
+import { TRIGGER_PRICE_TYPES } from "../../common.mjs";
```
Comment (lcaresia) [inline components/bybit/actions/linear-market-query-kline/linear-market-query-kline.mjs:2]: Move this to a `constants` file, will be more clear!

```suggestion
import { TRIGGER_PRICE_TYPES } from "../../common/constants.mjs";
```

File: components/bybit/actions/linear-order-cancel-all/linear-order-cancel-all.mjs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    bybit,
+    symbol: {
+      propDefinition: [
+        bybit,
+        "symbol",
+      ],
+    },
+  },
+  type: "action",
+  methods: {},
```
Comment (lcaresia) [inline components/bybit/actions/linear-order-cancel-all/linear-order-cancel-all.mjs:19]: You can remove empty `methods` from other files too.

```suggestion
```

File: components/bybit/actions/linear-stop-order-replace/linear-stop-order-replace.mjs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        "sl_trigger_by",
+      ],
+    },
+  },
+  type: "action",
+  methods: {},
+  async run({ $ }) {
+    const API_METHOD = "POST";
+    const API_PATH = "/private/linear/stop-order/replace";
+    let returnValue = await this.bybit.makeRequest(API_METHOD, API_PATH, this);
```
Comment (lcaresia) [inline components/bybit/actions/linear-stop-order-replace/linear-stop-order-replace.mjs:77]: You can update this in other files too.

```suggestion
    const returnValue = await this.bybit.makeRequest(API_METHOD, API_PATH, this);
```

**Review by lcaresia — CHANGES_REQUESTED** [review]
I didn't found any `async options`, you have sure that you can't implement in any prop?

**Comment by trendoscope-algorithms** [comment]
> I didn't found any `async options`, you have sure that you can't implement in any prop?

I have used options and Async options for many parameters. They are all in the app file.

File: components/bybit/actions/linear-order-cancel-all/linear-order-cancel-all.mjs (hunk 3 above)
Comment (trendoscope-algorithms) [reply components/bybit/actions/linear-order-cancel-all/linear-order-cancel-all.mjs]: Done

File: components/bybit/actions/linear-market-query-kline/linear-market-query-kline.mjs (hunk 2 above)
Comment (trendoscope-algorithms) [reply components/bybit/actions/linear-market-query-kline/linear-market-query-kline.mjs]: Done

File: components/bybit/actions/linear-stop-order-replace/linear-stop-order-replace.mjs (hunk 4 above)
Comment (trendoscope-algorithms) [reply components/bybit/actions/linear-stop-order-replace/linear-stop-order-replace.mjs]: Done

File: components/bybit/actions/generic-market-account-ratio/generic-market-account-ratio.mjs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  type: "action",
+  methods: {},
+  async run({ $ }) {
+    const API_METHOD = "GET";
+    const API_PATH = "/v2/public/account-ratio";
+    let returnValue = await this.bybit.makeRequest(API_METHOD, API_PATH, this);
+    if (returnValue.ret_code) {
+      throw new Error(returnValue.ret_msg);
+    } else {
+      $.export("$summary", "Long/Short Ratio Request Successful");
```
[the remark this answers is not in the document]
Comment (trendoscope-algorithms) [reply components/bybit/actions/generic-market-account-ratio/generic-market-account-ratio.mjs:40]: @lcaresia - Apart from return code and message, there is no generic info. 

image

Should I just print them?

[the remark this answers is not in the document]
Comment (lcaresia) [reply components/bybit/actions/generic-market-account-ratio/generic-market-account-ratio.mjs:40]: If the request don't return any ID, you don't need to put in the `export` message

**Review by lcaresia — APPROVED** [review]
LGTM!

**Comment by vunguyenhung** [comment]
Hi everyone, all test cases are passed! Ready for release! 

Test report
[Bybit_4456.pdf](https://github.com/PipedreamHQ/pipedream/files/9777718/Bybit_4456.pdf)

**Comment by vunguyenhung** [comment]
I will merge this on behave of @trendoscope-algorithms 🙏

**Comment by trendoscope-algorithms** [comment]
Thanks very much guys @feyzullah , @lcaresia , @vunguyenhung

**Pull request merged**